consensus/src/experimental/buffer_item.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::state_replication::StateComputerCommitCallBackType;
+use consensus_types::{common::Author, executed_block::ExecutedBlock};
+use diem_crypto::ed25519::Ed25519Signature;
+use diem_types::{
+    account_address::AccountAddress,
+    block_info::BlockInfo,
+    ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
+    validator_verifier::ValidatorVerifier,
+};
+use itertools::zip;
+use std::collections::BTreeMap;
+
+// we differentiate buffer items at different stages
+// for better code readability
+pub struct OrderedBufferItem {
+    pub pending_votes: BTreeMap<AccountAddress, Ed25519Signature>,
+    pub callback: StateComputerCommitCallBackType,
+    pub ordered_blocks: Vec<ExecutedBlock>,
+    pub ordered_proof: LedgerInfoWithSignatures,
+}
+
+pub struct ExecutedBufferItem {
+    pub executed_blocks: Vec<ExecutedBlock>,
+    pub pending_votes: BTreeMap<AccountAddress, Ed25519Signature>,
+    pub callback: StateComputerCommitCallBackType,
+    pub commit_info: BlockInfo,
+    pub ordered_proof: LedgerInfoWithSignatures,
+}
+
+pub struct SignedBufferItem {
+    pub executed_blocks: Vec<ExecutedBlock>,
+    pub commit_proof: LedgerInfoWithSignatures,
+    pub callback: StateComputerCommitCallBackType,
+}
+
+pub struct AggregatedBufferItem {
+    pub executed_blocks: Vec<ExecutedBlock>,
+    pub aggregated_proof: LedgerInfoWithSignatures,
+    pub callback: StateComputerCommitCallBackType,
+}
+
+pub enum BufferItem {
+    Ordered(Box<OrderedBufferItem>),
+    Executed(Box<ExecutedBufferItem>),
+    Signed(Box<SignedBufferItem>),
+    Aggregated(Box<AggregatedBufferItem>),
+}
+
+impl BufferItem {
+    pub fn new_ordered(
+        ordered_blocks: Vec<ExecutedBlock>,
+        ordered_proof: LedgerInfoWithSignatures,
+        callback: StateComputerCommitCallBackType,
+    ) -> Self {
+        Self::Ordered(Box::new(OrderedBufferItem {
+            pending_votes: BTreeMap::<AccountAddress, Ed25519Signature>::new(),
+            callback,
+            ordered_blocks,
+            ordered_proof,
+        }))
+    }
+
+    // pipeline functions
+    pub fn advance_to_executed(self, executed_blocks: Vec<ExecutedBlock>) -> Self {
+        match self {
+            Self::Ordered(ordered_item_box) => {
+                let ordered_item = *ordered_item_box;
+                assert_eq!(ordered_item.ordered_blocks.len(), executed_blocks.len());
```

Comment:
we can use zip_eq in the for loop to achieve this

Reply:
noted on TODO